### PR TITLE
device removal spec + implementation

### PR DESCRIPTION
**Pull request by mchalski**
Issues: MEN-985

Signed-off-by: mchalczynski <marcin.chalczynski@rndity.com>

@mendersoftware/rndity @maciejmrowiec

### CODE REVIEW

**Comment by mchalski** [comment]
ping - stacke the implementation task on top of this PR (MEN-987)

**Comment by maciejmrowiec** [comment]
quick question, should delete be part of internal or management?

**Comment by mchalski** [comment]
@maciejmrowiec hmm, now that you mention it, I think `internal` is the better choice.

this will take part in device decommissioning only, can't  see other use cases - will fix this

**Comment by mchalski** [comment]
@maciejmrowiec fixed

File: datastore_mongo.go [hunk 1]
```diff
@@ -85,6 +85,23 @@ func (db *DataStoreMongo) GetDevice(id DeviceID) (*Device, error) {
 	return &res, nil
 }
 
+func (db *DataStoreMongo) DeleteDevice(id DeviceID) error {
+	s := db.session.Copy()
+	defer s.Close()
+
+	filter := bson.M{"id": id}
```
Comment (maciejmrowiec) [inline datastore_mongo.go]: id or _id ?

Comment (maciejmrowiec) [inline datastore_mongo.go]: just browsed structs, it's `id`

Comment (mchalski) [inline datastore_mongo.go]: this time it's `id`. I think we started to use `_id` after deviceadm:

```
// Device wrapper                            
type Device struct {                         
    //system-generated device ID             
    ID DeviceID `json:"id" bson:",omitempty"`
...
```

Comment (bboozzoo) [inline datastore_mongo.go]: Very likely. 

BTW. on one hand there's a default index on `_id`. On the other, bson tag `_id` has some implicit, magical meaning. I guess we can accept the latter and just update the rest of the code to use `_id` consistently.

Comment (maciejmrowiec) [inline datastore_mongo.go]: If we change this now from `id` to `_id` this means we have to write migration on existing installations or support both at the same time. Not sure if worth the effort.

Comment (mchalski) [inline datastore_mongo.go]: no strong opinion, @maciejmrowiec what are we doing?
- leave as is
- migrate to `_id` in this task
- have another task for this

Comment (maciejmrowiec) [inline datastore_mongo.go]: I'm for keeping it as it is, if we require unique index (do we?) we can add extra one.

Comment (mchalski) [inline datastore_mongo.go]: not sure, but we use upsert, and there's this bit of warning in mongo docs:

https://docs.mongodb.com/manual/reference/method/db.collection.update/#use-unique-indexes

so I'm leaning towards 'yes we do'. thoughts? plus - where to put it - just stick it into code, or actually create a migration?

Comment (maciejmrowiec) [inline datastore_mongo.go]: Run ensure index on application startup (then initing db object for example) with building index in foreground to make sure we start application with existing index.

https://github.com/mendersoftware/deployments/blob/master/resources/images/mongo/images.go#L64

Comment (mchalski) [inline datastore_mongo.go]: stacked two commits on top:
- one with the index
- another with a possible travis fix (suggested by GregD) - the build keeps failing, same for devauth @bboozzoo

**Comment by mchalski** [comment]
great, now the coverage went down because of the new indexing method...guess no other useful work will be done today ;)

**Comment by mchalski** [comment]
ok, build is green, ready for merging

**Pull request merged**